src/main/kotlin/org/jitsi/nlj/transform/node/AudioRedHandler.kt

```diff
@@ -0,0 +1,267 @@
+    val distance = Config.config.distance
+    val vadOnly = Config.config.vadOnly
+
+    val stats = Stats()
+
+    var audioLevelExtId: Int? = null
+    var redPayloadType: Int? = null
+
+    private val ssrcRedHandlers: MutableMap<Long, SsrcRedHandler> = HashMap()
+
+    // TODO: Hook up to the current packet loss.
+    private fun currentLoss(): Double = 0.0
+
+    init {
+        streamInformationStore.onRtpPayloadTypesChanged { payloadTypes ->
+            redPayloadType = payloadTypes.values.find { it is AudioRedPayloadType }?.pt?.toInt()
+        }
+        streamInformationStore.onRtpExtensionMapping(RtpExtensionType.SSRC_AUDIO_LEVEL) {
+            audioLevelExtId = it
+        }
+    }
+
+    override fun transform(packetInfo: PacketInfo): List<PacketInfo> {
+        val audioPacket = packetInfo.packet as? AudioRtpPacket ?: return listOf(packetInfo)
+
+        val ssrcHandler = ssrcRedHandlers.computeIfAbsent(audioPacket.ssrc) { SsrcRedHandler() }
+
+        return when (audioPacket) {
+            is RedAudioRtpPacket -> ssrcHandler.transformRed(packetInfo)
+            else -> listOf(ssrcHandler.transformAudio(packetInfo))
+        }
+    }
+
+    override fun trace(f: () -> Unit) = f.invoke()
+
+    override fun getNodeStats(): NodeStatsBlock = super.getNodeStats().apply {
+        addString("red_payload_type", redPayloadType?.toString() ?: "null")
+        addString("audio_level_ext_id", audioLevelExtId?.toString() ?: "null")
+        addString("policy", policy.toString())
+        addString("distance", distance.toString())
+        addBoolean("vad_only", vadOnly)
+        addString("loss_threshold", lossThreshold.toString())
+
+        addNumber("red_packets_decapsulated", stats.redPacketsDecapsulated)
+        addNumber("red_packets_forwarded", stats.redPacketsForwarded)
+        addNumber("audio_packets_encapsulated", stats.audioPacketsEncapsulated)
+        addNumber("audio_packets_forwarded", stats.audioPacketsForwarded)
+        addNumber("lost_packets_recovered", stats.lostPacketsRecovered)
+        addNumber("redundancy_packets_added", stats.redundancyPacketsAdded)
+    }
+
+    /**
+     * Handler for a specific stream (SSRC)
+     */
+    private inner class SsrcRedHandler {
+        val sentAudioCache = RtpPacketCache(20)
+
+        /**
+         * Process an incoming audio packet. It is either forwarded as it is, or encapsulated in RED with previous
+         * packets optionally added as redundancy.
+         */
+        fun transformAudio(packetInfo: PacketInfo): PacketInfo {
+            // Whether we need to add RED encapsulation
+            val redPayloadType = redPayloadType
+            val encapsulate = when (redPayloadType) {
+                null -> false
+                else -> when (policy) {
+                    RedPolicy.NOOP, RedPolicy.STRIP -> false
+                    RedPolicy.PROTECT_ALL -> true
+                    // RedPolicy.PROTECT_DOMINANT -> =isDominant
```

Comment:
Are you planning to implement this?

Reply:
I don't know yet (I want to look at the data to find out how there are non-dominant speakers that send non-silence). I can remove it for now if you think it's only confusing.